libraries/AP_Compass/AP_Compass_IST8308.cpp

```diff
@@ -0,0 +1,246 @@
+#define CNTL2_VAL_CONT_ODR20_MODE   0x4
+#define CNTL2_VAL_CONT_ODR50_MODE   0x6
+#define CNTL2_VAL_CONT_ODR100_MODE  0x8
+#define CNTL2_VAL_CONT_ODR200_MODE  0xA
+#define CNTL2_VAL_CONT_ODR8_MODE    0xB
+#define CNTL2_VAL_CONT_ODR1_MODE    0xC
+#define CNTL2_VAL_CONT_ODR0P5_MODE  0xD
+#define CNTL2_VAL_SINGLE_TEST_MODE  0x10
+
+#define CNTL3_REG 0x32
+#define CNTL3_VAL_SRST 1
+#define CNTL3_VAL_DRDY_POLARITY_HIGH (1 << 2)
+#define CNTL3_VAL_DRDY_EN (1 << 3)
+
+#define CNTL4_REG 0x34
+#define CNTL4_VAL_DYNAMIC_RANGE_500 0
+#define CNTL4_VAL_DYNAMIC_RANGE_200 0x1
+
+#define OSRCNTL_REG 0x41
+#define OSRCNTL_VAL_XZ_1  (0)
+#define OSRCNTL_VAL_XZ_2  (1)
+#define OSRCNTL_VAL_XZ_4  (2)
+#define OSRCNTL_VAL_XZ_8  (3)
+#define OSRCNTL_VAL_XZ_16 (4)
+#define OSRCNTL_VAL_XZ_32 (4)
+#define OSRCNTL_VAL_Y_1  (0 << 3)
+#define OSRCNTL_VAL_Y_2  (1 << 3)
+#define OSRCNTL_VAL_Y_4  (2 << 3)
+#define OSRCNTL_VAL_Y_8  (3 << 3)
+#define OSRCNTL_VAL_Y_16 (4 << 3)
+#define OSRCNTL_VAL_Y_32 (5 << 3)
+
+#define SAMPLING_PERIOD_USEC (10 * AP_USEC_PER_MSEC)
+
+extern const AP_HAL::HAL &hal;
+
+AP_Compass_Backend *AP_Compass_IST8308::probe(AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                                              bool force_external,
+                                              enum Rotation rotation)
+{
+    if (!dev) {
+        return nullptr;
+    }
+
+    AP_Compass_IST8308 *sensor = new AP_Compass_IST8308(std::move(dev), force_external, rotation);
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_IST8308::AP_Compass_IST8308(AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                       bool force_external,
+                                       enum Rotation rotation)
+    : _dev(std::move(dev))
+    , _rotation(rotation)
+    , _force_external(force_external)
+{
+}
+
+bool AP_Compass_IST8308::init()
+{
+    uint8_t reset_count = 0;
+
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        return false;
+    }
+
+    // high retries for init
+    _dev->set_retries(10);
+
+    uint8_t whoami;
+    if (!_dev->read_registers(WAI_REG, &whoami, 1) ||
+        whoami != DEVICE_ID) {
+        // not an IST8308
+        goto fail;
+    }
+
+    for (; reset_count < 5; reset_count++) {
+        if (!_dev->write_register(CNTL3_REG, CNTL3_VAL_SRST)) {
+            hal.scheduler->delay(10);
+            continue;
+        }
+
+        hal.scheduler->delay(20);
+
+        uint8_t cntl3 = 0xFF;
+        if (_dev->read_registers(CNTL3_REG, &cntl3, 1) &&
+            (cntl3 & 0x01) == 0) {
+            break;
+        }
+    }
+
+    if (reset_count == 5) {
+        printf("IST8308: failed to reset device\n");
+        goto fail;
+    }
+
+    // DRDY enabled
+    // Dynamic Range=±500 uT, Sensitivity=6.6 LSB/uT
+    // OSR 16 (max 100Hz)
+    // Start continuous mode at 100Hz
+    if (!_dev->write_register(CNTL3_REG, CNTL3_VAL_DRDY_EN) ||
```

Comment:
Even after reading the doc for checked_registers, I miss a little bit how they work. How would the sensor recover? It seems it only writes back the value to that register, but those initialization registers can't be written to at any time, only during the initialization.

So... if they change during runtime after everything is initialized, it's probably a bus error (so writing it back doesn't help),  or it went so bad that we should go through the entire initialization again (or just mark the sensor as unhealthy).

Reply:
It is sensor dependent. Many (most?) sensors do work if you re-write the config values after power on. One of the situations it handles is a temporary power outage on the external I2C bus. We have had that happen in real flights, and the checked register support does save aircraft sometimes. We've also had sensors suddenly change config, perhaps due to corrupted bus transfer. The checked register code also helps there.
Whether it is applicable to a particular sensor needs to be checked by the driver author. I often test this by unplugging a sensor while its running then plugging it back in again, and checking it re-initialises properly. Or I use devop in mavproxy to deliberately corrupt a sensors registers while running, and check that it re-inits correctly.

Reply:
ok, but then let's say the register for sampling frequency changes. I can't simply write it back to the previous value without also checking if the mode because I can't change the frequency without changing the mode back to standby. And if the mode change, we better also configure the frequency.   If power is removed, the sensor goes back to the default values, which basically means we'd have to go through the initialization again.  A nice thing about the continuous mode (other than the much better timing of the samples) is that after initialization we don't write to it anymore, it's only read operations.

Reply:
that is why check_next_register() returns a bool, so the driver can do a complete new setup if it wants to if the checker does find the registers have unexpectedly changed

Reply:
Sounds good, but can be done on top, ok?

Reply:
sure